pkg/access_token/access_token_handler.go

```diff
@@ -0,0 +1,7 @@
+package access_token
+
+type AccessTokenHandler interface {
```

Comment:
1. It's not clear to me why AccessTokenHandler does from its name. It would be useful to have a brief description of AccessTokenHandler for godocs.
2. Reading this PR it's unclear why Read and Delete are necessary. I haven't seen them be exercised outside of the unit tests for a given AccessTokenHandler implementation.

Reply:
Well, actually they are not used here. they are used in the project that will consume this authenticator. The interface needs to be here so objects can be created in the consuming project. 

For example, in the `conjur-k8s-secrets-manager` we will use `AccessTokenMemory`. we will create the `authenticator` object with this handler which will not write the data to a file. later on, we will use the `Read()` method to get the token for retrieving secrets from conjur.

developers who consume this project can create their own `AccessTokenHandler` but these two are used by our projects so they are fully integrated in the project. think about it like with ansible. you can create your own plugin and if it's good enough it will be implemented inside the ansible project.

Reply:
> It's not clear to me why AccessTokenHandler does from its name.

isn;t it clear from its methods? not sure what i can write besides 
```
This interface handles a Conjur access token. Classes implementing this interface should have the ability to read, write & delete the access token
```

Reply:
Adding examples, as you did in your first response comment, makes it all the more concrete. It's this sort of description which would be useful for someone to grok the utility of the interface - the rich context that comes from examples is invaluable.

Reply:
cool, i added a description

Reply:
@orenbm Thanks for including the description. I was thinking we could provide a description that alludes to example usage of the methods e.g. `conjur-k8s-secrets-manager` or Secretless. When I initially thought about the example I was looking towards this:

> For example, in the conjur-k8s-secrets-manager we will use AccessTokenMemory. we will create the authenticator object with this handler which will not write the data to a file. later on, we will use the Read() method to get the token for retrieving secrets from conjur.

Reply:
I'm curious why we don't just have an object `accessToken` that has methods on it `read`, `write`, `delete`. what is the value of separating this out into a "handler"? What is the purpose of having a separate "memory" object?

Reply:
@izgeri just to make sure i understand you, are you suggesting that we have an `AccessToken` object which will hold the logic of wether we should write the token to a file or to the memory? This means that the `AccessToken` needs to be aware of the use-case we're at and will have some `if` clause or switch case in it.

If that's what you mean then i think it is better to have different structs that implement the `handler` interface and have the logic outside. the word `handler` is just for readability as i think it better explains what the object does (it handles an access token it's not an actual access token). 

what do you think?

Reply:
Consider the following a fresh comment: 

My big comment is that the interface currently called `AccessTokenHandler` is actually just defining the concept of an access token.  It should be called `AccessToken`.  Then file and memory are two specific implementation of an access token (re: the conversation above with Geri, I am agnostic as to whether or not having those to implementations is a good idea, that all depends on what you're doing with them).

But assuming that it is a good idea, the package design convention you should follow is to have a structure like this:
```
access_token.go (defines the interface only)
    file
        access_token.go (file impl in file package.  client code, eg, creates one like  file.NewAccessToken)
    memory
        access_token.go (memory imple.  client code does memory.NewAccessTokent(...), etc)
```

So, 

1. root `accesstoken` package defines the interface only and has no dependencies.
2. packaces below that `file` and `memory` define the specific implementations and do have deps.
3. Note that this cleans up the naming as client code does `file.NewAccessToken` instead of `NewAccessTokenFile`.

Reply:
nice!

Reply:
done refactor

Reply:
Noice!!  You don't mess around.

One small leftover thing:

```
type AccessTokenFile struct {
	Data          []byte
	TokenFilePath string
}
```
can now become `AccessToken` because you're already in the `file` package now.  This is called "package name stutter" and is a common mistake to be avoided.

I might consider changing `TokenFilePath` to `PathToToken` but that's more of a subjective personal suggestion: take it or leave it.

Reply:
done, thanks

Reply:
I spent some time looking at this today, and I have other notes about the code base I'll try to post later, but re: the `file.AccessToken` specifically, I think a couple things struck me as problematic.

1. Afaict, you must call `Write()` before you call `Read()`.  So clients of this object cannot simply create an object and then use it.  They must create the object, _and then_ call `Write()` before the object is really "ready".  Thus the construction responsibility has been split up across two methods, rather than residing in the constructor.  A fun old-school term for this kind of coupling is [Connascence of Execution](https://connascence.io/execution.html).
2. When you `Read()` back the value, you read it from memory.  Indeed, it errors if the token is not in memory:

   ```go
   if token.Data == nil {
       return nil, fmt.Errorf("error reading access token, reason: data is empty")
   }
   ```
   So really, this `file.AccessToken` appears to be fundamentally the same as the `memory.AccessToken`, except when you `Write()` it has the _side effect of creating a file_.  However, I wasn't able to find in the code where this file is ever read after being created.  I may have missed it -- if so, where does this happen?   In any case, the reason for this design I assume is so that in memory data and the file can travel through the program as a single unit, and both will be cleaned up together when `Delete()` is called.  However, I think I'd prefer a different higher-level code structure that made the use the file clearer.  But that would probably require larger changes the code, and I'd want to understand it better before suggesting anything.

Reply:
@jonahx hi, thanks for the detailed comment :)
1. You right, there is a  construction responsibility. 
2. The file is never read after bean created, but it should not deleted from the file system, when we are in SECRET_DESTINATION=none mode, for the application to use it to fetch secrets.

i would like to here how do you suggest to redesign it  ?

Reply:
> when we are in SECRET_DESTINATION=none mode, for the application to use it to fetch secrets.